design-proposals/shadow-node.md

```diff
@@ -0,0 +1,352 @@
+  |-----------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+  | object.spec == oldObject.spec                                                                                                     | virt-handler user cannot modify spec of the nodes                              |                                                                                                                                                                        |
+  | object.status == oldObject.status                                                                                                 | virt-handler user cannot modify status of the nodes                            |                                                                                                                                                                        |
+  | object.(not labels/annotations/resourceVersion/managedFields) == oldObject.(not labels/annotations/resourceVersion/managedFields) | virt-handler user can only change allowed sub-metadata fields.                 |                                                                                                                                                                        |
+  | size(`newNonKubevirtLabels`) == size(`oldNonKubevirtLabels`)                                                                      | virt-handler user cannot add/delete non kubevirt-owned labels                  |                                                                                                                                                                        |
+  | for label in `newNonKubevirtLabels`: {(label in `oldNonKubevirtLabels`) AND `newLabels`[label] == `oldLabels`[label]}             | virt-handler user cannot update non kubevirt-owned labels                      |                                                                                                                                                                        |
+
+### Backwards compatibility:
+validationAdmissionPolicy [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it should be OK for deployment on Kubevirt main and next minor release branches.
+However, since kubevirt needs to be supported up to 2 k8s versions back, then the deployment of these new objects needs to be conditioned by the k8s version deployed. 
+The virt-operator deployment logic should be: 
+```
+if k8s-ver >= v1.30 
+then
+  deploy the objects
+else
+  deploy, but only log on failure
+```
+Once k8s v1.29 and below is not supported on kubevirt, this logic should be removed.
```

Comment:
What if Kubevirt is being installed on k8s 1.29 (where validationAdmissionPolicy is beta), but the feature gate is enabled? I think it's reasonable to deploy the objects on such scenario, and it might even be an acceptable workaround for Kubevirt users until they're able to upgrade thier k8s version.

Perhaps we can take a different approach of using dry-run to determine if the object is known or not.
That is:
* Create the objects with a dry-run option. Check if an error is returned.
  * If a nil error is returned - re-create without the dry-run option.
  * If an "object unknown" error is returned ([this one](https://github.com/kubernetes/kube-openapi/blob/f0e62f92d13f418e2732b21c952fd17cab771c75/pkg/util/proto/validation/errors.go#L78)?), log and continue.
  * If a different error is returned - fail.

We can obviously drop the second step once we drop support for k8s 1.29.

Reply:
Hey, I'm not sure why we need to dry run - if the object is not installed it'll not deploy anyway right? . 

But we do want to distinguish two behaviors:
For 1.29-, we want to deploy (or at least try), if we get error, log it.
If it's 1.30+, we do want to fail if errors, and not only log.

Reply:
> Ignoring an "object not found" error in the general case is wrong imo

I agree, but relying on k8s version is (as someone also [said here](https://github.com/kubevirt/community/pull/246#discussion_r1586582058) :wink:) also "unusual" :) The question is what's worse.

But TBH I initially thought that for 1.29- we're going to not even to deploy the objects, which sounded problematic. If you're going to simply try to create the objects on a best-effort basis and log on error it sounds better to me.

Thanks!